appcues/src/main/java/com/appcues/analytics/ExperienceLifecycleTracker.kt

```diff
@@ -20,29 +23,28 @@ import com.appcues.statemachine.StateMachine
 import kotlinx.coroutines.CoroutineDispatcher
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.launch
-import kotlinx.coroutines.withContext
 import org.koin.core.component.KoinScopeComponent
 import org.koin.core.component.inject
 import org.koin.core.scope.Scope
 import java.util.Date
 
 internal class ExperienceLifecycleTracker(
-    override val scope: Scope
+    override val scope: Scope,
 ) : KoinScopeComponent {
 
     // lazy property injection to avoid circular DI reference in constructor
     // AnalyticsTracker -> this <- Analytics Tracker
     private val analyticsTracker: AnalyticsTracker by inject()
-    private val stateMachine: StateMachine by inject()
     private val storage: Storage by inject()
+    private val appcuesCoroutineScope: AppcuesCoroutineScope by inject()
 
-    suspend fun start(dispatcher: CoroutineDispatcher = Dispatchers.IO): Unit = withContext(dispatcher) {
-        launch { observeState() }
-        launch { observeErrors() }
+    fun start(stateMachine: StateMachine, dispatcher: CoroutineDispatcher = Dispatchers.IO) {
+        appcuesCoroutineScope.launch(dispatcher) { stateMachine.observeState() }
+        appcuesCoroutineScope.launch(dispatcher) { stateMachine.observeErrors() }
     }
 
-    private suspend fun observeState() {
-        stateMachine.stateFlow.collect {
+    private suspend fun StateMachine.observeState() {
+        stateFlow.collect {
```

Comment:
Is there a way to end this collection on the flow when the state machine completes? Wondering if this will leave any memory references alive indefinitely now that we have the potential for multiple state machines during the app lifetime.

I guess a similar situation/question exists for the AppcuesViewModel usage as well.

Reply:
yes once we get to the point were we want to clean state machines lets say on Screen calls, we can then store those launch into Job variables and cancel those whenever we want